### DISCUSSION TITLE
Trees of the ensemble present unnecessary splits

### DISCUSSION THREAD

**Discussion by lorebon** [wildboar-foundation/wildboar#87] [General]
Dear Isak,

during my experiments I've encountered some ShapeletTrees that present meaningless splits, in the sense that 2 leaf nodes deriving from a certain branching node have the same classification label.

Is there a reason for that? It seems that the tree wants to reach the max depth regardless. I'm sure this does not impact performance, but I still wonder why such behavior.

Thanks!

**Comment by isaksamsten** [comment]
It sounds like a bug in the stopping criteria. Any way for me to reproduce?

**Comment by isaksamsten** [comment]
Could you try setting `min_impurity_decrease` to a value slightly larger than 0 and see if the issue persists? Could be some minor numerical deviation where the impurity should be zero but is very slightly above du to numerical instability.

**Reply by lorebon** [reply]
Ok I'm trying with min_impurity_decrease equal to 1e-3 and it seems working! Thank you!

**Reply by isaksamsten** [reply]
Thanks for the report. I will push a small change to fix the issue.

**Comment by lorebon** [comment]
I'm still facing the same issue with a value of 0.08. 

Here is the minimal reproducible example I can provide:

[test_min_decrease.txt](https://github.com/wildboar-foundation/wildboar/files/13311107/test_min_decrease.txt)

If you run it you can notice that leaf nodes 4 and 5 have the same label (1).

With min_impurity_decrease above 0.08 there is only the root node.

**Reply by isaksamsten** [reply]
The prediction in node `3` is `[0.5, 0.5]` and `np.argmax` simply return the first so I don't think this example reproduces the problem. Could you check that the offending nodes doesn't have the same probabilities?

**Reply by isaksamsten** [reply]
Here is the code I used:

```python
from wildboar.datasets import load_dataset
from wildboar.ensemble import ShapeletForestClassifier

X_train, X_test, y_train, y_test = load_dataset("Lightning2", preprocess="minmax_scale", merge_train_test=False)
X_train = np.round(X_train, 3)
J = X_train.shape[1]
H = 170

clf = ShapeletForestClassifier(
    n_estimators=1,
    max_depth=3,
    min_shapelet_size=H / J,
    max_shapelet_size=H / J,
    min_impurity_decrease=0.08,
    metric="manhattan",
    random_state=1,
)
clf.fit(X_train, y_train)
tree = clf.estimators_[0].tree_
labels = np.full(tree.node_count, np.nan)
temp = clf.classes_.take(np.argmax(tree.value[tree.left == -1], axis=1))
labels[tree.left == -1] = temp
print(labels)
print(tree.value)
```

**Reply by lorebon** [reply]
What does the 'value' attribute of 'tree' encode? 

Nonetheless, I think the problem lies in np.argmax taking the first value in case of ties. So a solution could be just enforcing different labels in such case. Thank you!

**Reply by isaksamsten** [reply]
The attribute `value` encodes the probability distribution over the labels reaching the node. So in this case 18 samples reaches the node (you can use the attribute `n_node_samples` to confirm this) out of which 8 are labeled as -1 and 8 as 1.

**Reply by lorebon** [reply]
I've just checked and nodes 4 and 5 present the following probabilities:
4) 0.5333 0.4666
5) 0.8857 0.1143
So the np.argmax is rightfully choosing label 1 for both nodes.

**Reply by isaksamsten** [reply]
What version are you using? Just so I test the right version

**Reply by lorebon** [reply]
I tried with both 1.1.2 and 1.1.3

**Reply by isaksamsten** [reply]
I have a hypothesis on what can go wrong. [here](https://github.com/wildboar-foundation/wildboar/blob/5d97ac139eed6e44077dcbca31aef40c80b04745/src/wildboar/tree/_ctree.pyx#L1096) we skip shapelets that produce constant distances to all samples. The bug would present it self if all shapelets produce such constant distances and the best split is not initialised. Perhaps that’s what’s happening. I will try to implement some sanity check for that case and see if I can reproduce the bug.

**Comment by isaksamsten** [comment]
Hi! I will take a look! Thanks for reporting
